### ISSUE TITLE
Liberty API/SPI javadoc needs building for Open Liberty

### ISSUE THREAD

**Issue by lauracowen** [OpenLiberty/open-liberty#14629]
## Description

The Open Liberty api/spi javadoc is currently generated and shipped with the product in the /wlp/dev/ folder, however it is not zipped up at the time of the build. This feature is to create a .jar with the  Open Liberty api/spi javadoc so that it can be displayed on the openliberty.io website.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## Documents

When available, add links to required feature documents. Use "N/A" to mark particular documents which are not required by the feature.

- Aha: Externally raised RFE ([Aha](https://cloud-platform.ideas.aha.io/))
  - Link the RFE with this issue
- UFO: Link to [Upcoming Feature Overview](https://ibm.box.com/v/UFO-Template) document
  - Set the Box link to be publicly accessible, with a long expiration (10 years)
    - Click "Share" > select "People with link" > click "Link Settings" > under "Link Expiration" select "Disable Shared Link on" > set an expiration date ~10 years into the future
    - If you lack permissions, contact [OpenLiberty/release-architect](https://github.com/orgs/OpenLiberty/teams/release-architect)
- FTS: [Link to Feature Test Summary GH Issue](https://github.com/OpenLiberty/open-liberty/issues/24886)
- Beta Blog: Link to Beta Blog Post GH Issue
- GA Blog: Link to GA Blog Post GH Issue

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## Process Overview

- [Prioritization](#prioritization)
- [Design](#design)
- [Implementation](#implementation)
- [Legal and Translation](#legal-and-translation)
- [Beta](#beta)
- [GA](#ga)
  - [Focal Point Approvals](#focal-point-approvals-complete-by-feature-complete-date)
- [Other Deliverables](#other-deliverables)

### General Instructions

The process steps occur roughly in the order as presented. Process steps occasionally overlap.

Each process step has a number of tasks which must be completed or must be marked as not applicable ("N/A").

Unless otherwise indicated, the tasks are the responsibility of the Feature Owner or a Delegate of the Feature Owner.

If you need assistance, reach out to the [OpenLiberty/release-architect](https://github.com/orgs/OpenLiberty/teams/release-architect).

**Important: Labels are used to trigger particular steps and must be added as indicated.** 

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Prioritization** (Complete Before Development Starts)

The ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) and area leads are responsible for prioritizing the features and determining which features are being actively worked on.

### **Prioritization**
- [ ] Feature added to the "New" column of the [Open Liberty project board](https://github.com/orgs/OpenLiberty/projects/2)
  - Epics can be added to the board in one of two ways:
    - From this issue, use the "Projects" section to select the appropriate project board.
    - From the appropriate project board click "Add card" and select your Feature Epic issue
- [ ] Priority assigned
  - Attend the Liberty Backlog Prioritization meeting

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Design** (Complete Before Development Starts)

Design preliminaries determine whether a formal design, which will be provided by an [Upcoming Feature Overview (UFO)](https://ibm.box.com/v/UFO-Template) document, must be created and reviewed.  A formal design is required if the feature requires any of the following: UI, Serviceability, SVT, Performance testing, or non-trivial documentation/ID.

### **Design Preliminaries**
- [ ] UI requirements identified. (Owner and [UI focal point](https://github.com/orgs/OpenLiberty/teams/ui-approvers))
- [ ] ID requirements identified. (Owner and [ID focal point](https://github.com/orgs/OpenLiberty/teams/id-approvers))
   - Refer to [Documenting Open Liberty](https://github.com/OpenLiberty/open-liberty/wiki/Documenting-Open-Liberty).
   - Feature Owner adds label `ID Required`, if non-trivial documentation needs to be created by the ID team.
   - ID adds label `ID Required - Trivial`, if no design will be performed and only trivial ID updates are needed.
- [ ] Serviceability Requirements Identified. (Owner and [Serviceability focal point](https://github.com/orgs/OpenLiberty/teams/serviceability-approvers))
- [ ] SVT Requirements Identified. (Owner and [SVT focal point](https://github.com/orgs/OpenLiberty/teams/svt-approvers))
- [ ] Performance testing requirements identified. (Owner and [Performance focal point](https://github.com/orgs/OpenLiberty/teams/performance-approvers))

### **Design**
- [ ] POC Design / UFO review requested.
  - Owner adds label `Design Review Request`
- [ ] POC Design / UFO review scheduled.
  - Follow the instructions in POC-Forum repo
- [ ] POC Design / UFO review completed.
- [ ] POC / UFO Review follow-ons completed.
- [ ] POC Design / UFO approval requested.
  - Owner adds label `Design Approval Request`
- [ ] Design / UFO approved. ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) or N/A
  - ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) adds label `Design Approved`
  - Add the public link to the UFO in Box to the [Documents](#documents) section.
  - The UFO must always accurately reflect the final implementation of the feature. Any changes must be first approved. Afterwards, update the UFO by creating a copy of the original approved slide(s) at the end of the deck and prepend "OLD" to the title(s). A single updated copy of the slide(s) should take the original's place, and have its title(s) prepended with "UPDATED".

### **No Design**
- [x] No Design requested.
  - Owner adds label `No Design Approval Request`
- [ ] No Design / No UFO approved. ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) or N/A
  - Approver adds label `No Design Approved`

### **FAT Documentation**
- [ ] "Feature Test Summary" child task created
  - Use the [Feature Test Summary Template](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=&labels=Feature+Test+Summary&template=feature_test_summary.md&title=)
  - Add FTS issue link to the [Documents](#documents) section.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Implementation**

A feature must be [prioritized](https://github.com/orgs/OpenLiberty/projects/2) before any implementation work may begin to be delivered (inaccessible/no-ship).  However, a design focused approach should still be applied to features, and developers should think about the feature design prior to writing and delivering any code.  
Besides being prioritized, a feature must also be socialized (or No Design Approved) before any beta code may be delivered.  All new Liberty content must be inaccessible in our GA releases until it is [Feature Complete](#feature-complete) by either marking it `kind=noship` or [beta fencing](#beta-code) it.  
Code may not GA until this feature has obtained the "Design Approved" or "No Design Approved" label, along with all other tasks outlined in the [GA](#ga) section.

### **Feature Development Begins**
- [ ] Add the `In Progress` label

## **Legal and Translation**

In order to avoid last minute blockers and significant disruptions to the feature, the **legal items need to be done as early in the feature process as possible**, either in design or as early into the development as possible.  Similarly, translation is to be done concurrently with development.  Both **MUST** be completed before Beta or GA is requested.

### **Legal** (Complete before Feature Complete Date)
- [ ] Changed or new open source libraries are cleared and approved, or N/A. (Legal Release Services/Cass Tucker/Release PM).
- [ ] Licenses and Certificates of Originality (COOs) are updated, or N/A

### **Translation** (Complete 1 week before Feature Complete Date)
- [ ] PII updates are merged, or N/A. Note timing with translation shipments.

### **Innovation** (Complete 1 week before Feature Complete Date)
- [ ] Consider whether any aspects of the feature may be patentable. If any identified, disclosures have been submitted.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Beta**

In order to facilitate early feedback from users, all new features and functionality should first be released as part of a beta release.

### **Beta Code**
- [ ] Beta fence the functionality
  - `kind=beta`, `ibm:beta`, `ProductInfo.getBetaEdition()`
- [ ] Beta development complete and feature ready for inclusion in a beta release
  - Add label `target:beta` and the appropriate `target:YY00X-beta` (where YY00X is the targeted beta version).
- [ ] Feature delivered into beta
  - ([OpenLiberty/release-manager](https://github.com/orgs/OpenLiberty/teams/release-manager)) adds label `release:YY00X-beta` (where YY00X is the first beta version that included the functionality).

### **Beta Blog** (Complete 1.5 weeks before beta eGA)
- [ ] Beta blog issue created and populated using the [Open Liberty BETA blog post](https://github.com/OpenLiberty/open-liberty/issues/new/choose) template.
  - Add a link to the beta blog issue in the [Documents](#documents) section.
  - Note: This is for inclusion into the overall [beta release blog post](https://openliberty.io/blog/?search=beta&key=tag).  If, in addition, you'd also like to create a dedicated blog post about your feature, then follow the "Standalone Feature Blog Post" instructions under the [Other Deliverables](#other-deliverables) section.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **GA**

A feature is ready to GA after it is Feature Complete and has obtained all necessary Focal Point Approvals. 

### **Feature Complete**
- [ ] Feature implementation and tests completed.
  - [ ] All PRs are merged.
  - [ ] All epic and child issues are closed.
  - [ ] All stop ship issues are completed.
- [ ] Legal: all necessary approvals granted.
- [ ] Translation: All messages translated or sent for translation for upcoming release 
- [ ] GA development complete and feature ready for inclusion in a GA release
  - Add label `target:ga` and the appropriate `target:YY00X` (where YY00X is the targeted GA version).
  - Inclusion in a release requires the completion of all Focal Point Approvals.

### **Focal Point Approvals** (Complete by Feature Complete Date)

These occur only after GA of this feature is requested (by adding a `target:ga` label).  GA of this feature may not occur until all approvals are obtained.

### **All Features**
- [ ] **APIs/Externals** Externals have been reviewed or N/A.  ([OpenLiberty/externals-approvers](https://github.com/orgs/OpenLiberty/teams/externals-approvers))
  - Approver adds label `focalApproved:externals` 
- [ ] **Demo** Demo is scheduled for an upcoming EOI or N/A. ([OpenLiberty/demo-approvers](https://github.com/orgs/OpenLiberty/teams/demo-approvers))
  - Add comment `@OpenLiberty/demo-approvers Demo scheduled for EOI 23.7` to this issue.
  - Approver adds label `focalApproved:demo`.
- [ ] **FAT** All Tests complete and running successfully in SOE or N/A. ([OpenLiberty/fat-approvers](https://github.com/orgs/OpenLiberty/teams/fat-approvers))
  - Approver adds label `focalApproved:fat`.
- [ ] **Globalization** Translation and TVT are complete or N/A. ([OpenLiberty/globalization-approvers](https://github.com/orgs/OpenLiberty/teams/globalization-approvers))
  - Approver adds label `focalApproved:globalization`.

### **Design Approved Features**
- [ ] **Accessibility** Accessibility testing completed or N/A. ([OpenLiberty/accessibility-approvers](https://github.com/orgs/OpenLiberty/teams/accessibility-approvers))
  - Approver adds label `focalApproved:accessibility`.
- [ ] **ID** Documentation is complete or N/A. ([OpenLiberty/id-approvers](https://github.com/orgs/OpenLiberty/teams/id-approvers))
  - Approver adds label `focalApproved:id`.
  - > **_NOTE:_**  If only trivial documentation changes are required, you may reach out to the ID Feature Focal to request a `ID Required - Trivial` label.  Unlike features with regular ID requirement, those with `ID Required - Trivial` label do not have a hard requirement for a Design/UFO.
- [ ] **Performance** Performance testing is complete or N/A. ([OpenLiberty/performance-approvers](https://github.com/orgs/OpenLiberty/teams/performance-approvers))
  - Approver adds label `focalApproved:performance`.
- [ ] **Serviceability** Serviceability has been addressed or N/A. ([OpenLiberty/serviceability-approvers](https://github.com/orgs/OpenLiberty/teams/serviceability-approvers))
  - Approver adds label `focalApproved:sve`.
- [ ] **STE** Skills Transfer Education chart deck is complete or N/A. ([OpenLiberty/ste-approvers](https://github.com/orgs/OpenLiberty/teams/ste-approvers))
  - Approver adds label `focalApproved:ste`.
- [ ] **SVT** System Verification Test is complete or N/A. ([OpenLiberty/svt-approvers](https://github.com/orgs/OpenLiberty/teams/svt-approvers))
  - Approver adds label `focalApproved:svt`.

### **Remove Beta Fencing** (Complete by Feature Complete Date)
- [ ] Beta guards are removed, or N/A
  - Only after all necessary Focal Point Approvals have been granted.

### **GA Blog** (Complete by Feature Complete Date)
- [ ] GA Blog issue created and populated using the [Open Liberty GA release blog post](https://github.com/OpenLiberty/open-liberty/issues/new/choose) template.
  - Add a link to the GA Blog issue in the [Documents](#documents) section.
  - Note: This is for inclusion into the overall [release blog post](https://openliberty.io/blog/?search=release&key=tag).  If, in addition, you'd also like to create a dedicated blog post about your feature, then follow the "Standalone Feature Blog Post" instructions under the [Other Deliverables](#other-deliverables) section.

### **Post GA**
- [ ] Replace `target:YY00X` label with the appropriate `release:YY00X`. ([OpenLiberty/release-manager](https://github.com/orgs/OpenLiberty/teams/release-manager))

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Other Deliverables**

- [ ] **Standalone Feature Blog Post** A blog post specifically about your feature or N/A.  ([OpenLiberty/release-architect](https://github.com/orgs/OpenLiberty/teams/release-architect))
  - This should be strongly considered for larger or more prominent features.
  - Follow [instructions](https://github.com/OpenLiberty/blogs/tree/draft#writing-and-publishing-blog-posts-on-the-openlibertyio-blog) in the blogs repo.
- [ ] **OL Guides** OL Guides assessment is complete or N/A. ([OpenLiberty/guide-assessment](https://github.com/orgs/OpenLiberty/teams/guide-assessment/members))
- [ ] **Dev Experience** Developer Experience & Tools work is complete or N/A. ([OpenLiberty/dev-experience-assessment](https://github.com/orgs/OpenLiberty/teams/dev-experience-assessment/members))

Provide Javadoc for the Open Liberty API/SPI. There are two possible solutions:

1. Do all the API and SPI for Liberty together. This is different from how Javadoc is produced and shipped with Open Liberty, but more similar to what we do for MicroProfile and Jakarta EE
2. Do it feature by feature. This is the simplest solution. The downside is you don't have a central place showing all the Javadoc, but it makes it simple to tell what code is available per feature. Another pro for this design is when we do Jakarta EE 9 some of the packages will reference Jakarta EE 9 classes when enabled with Jakarta EE 9 features, but the Jakarata EE 8 features will reference Jakarta EE 8 classes so we will need to deal with two copies of Javadoc either way, and this is perhaps simpler.

Going to investigate option 1 as the preferred option.

**Comment by andymc12** [comment]
I'd like to contribute to this issue if possible.  Here are some ideas:
1) Keep just one "copy" of the OL API/SPI Javadocs (as opposed to a new copy for each release).  For one, the OL APIs/SPIs don't change that much, so multiple copies will just consume unnecessary space and make it more confusing for users. In order clarify that an API is new in a specific version of OL, we should be mandating the use of the `@since` javadoc tag - i.e. `@since 21.0.0.4`. 
2) Put all of the javadocs in a combined location - i.e. JAX-RS OL APIs will be listed in the same framesets as JPA OL APIs, etc. In order to tie the particular APIs back to the features or API/SPI jars that they are associated with, we should mandate that the `package-info.java` should contain that text - something like:
```
/**
 * ...
 * The APIs in this package are part of the following features:
 * <ul>
 * <li>jaxrs-2.0</li>
 * <li>jaxrs-2.1</li>
 * </ul>
 * <p>
 * In order to compile your applications when using these APIs, you will need to use the following Maven coordinates:
 * </p>
 * <pre>
 * <dependency>
 *    <groupId>com.ibm.websphere.appserver.api</groupId>
 *    <artifactId>com.ibm.websphere.appserver.api.jaxrs20</artifactId>
 *    <version>[1.0,2.0)</version>
 *    <scope>provided</scope>
 * </dependency>
 * </pre>
```
3) APIs should be separate from SPIs - both javadoc framesets should be "peers" to the existing Jakarta EE and MicroProfile javadoc framesets.

If we agree with this approach, we'll need to do the following:
1) Update the javadocs to include the public feature and maven coordinates. I think we should leave the `@since` tag off of existing APIs, but add them all new ones - starting with those added in #16069.
2) Add new build tooling to create a combined javadoc.
3) Each release, we'll need to compare the newly-released javadoc with what is currently published.  If anything has changed, we'll need to make an update.  We could do this as part of the release process, but ideally this would be automated - I'm thinking something like how dependabot works - it opens a PR with the updated javadoc, and then somebody from the release team just needs to merge it.

**Comment by mbroz2** [comment]
Regarding updating the API/SPI each release, one approach would be to place the API/SPI in a new repo.  This would allow us to simply force push the content to the repo each release, no need for PR, so no need for human approval/intervention.  The con being that it's yet another (doc) content repo for ol.io, and not part of the existing docs-javadoc repo (nor docs-generated).

Regarding Stage 3a, I'm wondering if there's already enough information today on the feature page to dynamically create links to the relevant javadoc page.  Since we know the packages each feature exposes and whether it's included in EE7, EE8, and/or MP, we may be able to dynamically create links to the package view in the javadoc.
For example, [the first package ("javax.batch.api") from the "Batch API 1.0" feature](https://www.openliberty.io/docs/21.0.0.3/reference/feature/batch-1.0.html#_standard_api_packages_provided_by_this_feature)
would link to https://www.openliberty.io/docs/21.0.0.3/reference/javadoc/liberty-javaee8-javadoc.html#class=javax/batch/api/package-summary.html&package=javax/batch/api/package-frame.html

Depending on how we want to implement things, we could then also link that package to EE7 javadoc (many packages also overlap with MP javadocs; for example, CDI).  Aka, we need a decision on whether we link to every javadoc, just one, or something in between.

**Comment by hlhoots** [comment]
website work issue to display OL api/spi is here: https://github.com/OpenLiberty/openliberty.io/issues/1550

**Comment by hlhoots** [comment]
issue to produce the jar for the website is here: https://github.ibm.com/websphere/WS-CD-Open/issues/27781

**Comment by hlhoots** [comment]
Latest status update on this item:  In September, we delivered a jar file which contains the javadoc similar to what is created for Knowledge Center, however the website needs a different format based on feedback from @kinueng last week.  Will need to work with Kin and figure out the correct format and make updates accordingly in the New Year.

**Comment by hlhoots** [comment]
Latest status Kin got back to me yesterday with details on how the MP javadoc is created.  The file delivered in September is the javadoc as it ships with the product in the /dev folder which is generated in frameless mode.  The website needs to use the frame version of the javadoc.  Working through a solution to provide that format.

**Comment by hlhoots** [comment]
Continuing to work on this with the website team to determine the proper format for display.

**Comment by hlhoots** [comment]
FTS issue: https://github.com/OpenLiberty/open-liberty/issues/24886

**Comment by hlhoots** [comment]
@OpenLiberty/demo-approvers Demo scheduled for EOI 23.7

**Comment by hlhoots** [comment]
The Open Liberty work is completed for this epic, and the remainder work of displaying the api / spi on the openliberty.io site can be found here https://github.com/OpenLiberty/openliberty.io/issues/1550
